cmd/src/headers.go

```diff
@@ -2,27 +2,51 @@ package main
 
 import (
 	"os"
+	"regexp"
 	"strings"
 )
 
-// parseAdditionalHeaders reads the environment for values like SRC_HEADER_NAME=VALUE
-// and creates a `{NAME: VALUE}` map. These headers should be applied to each request
-// to the Sourcegraph instance, as some private instances require special auth or
+// parseAdditionalHeaders reads the environment for values like SRC_HEADER_NAME=VALUE or
+// SRC_HEADERS and creates a `{NAME: VALUE}` map. These headers should be applied to each
+// request to the Sourcegraph instance, as some private instances require special auth or
 // additional proxy values to be passed along with each request.
 func parseAdditionalHeaders() map[string]string {
-	return parseAdditionalHeadersFromMap(os.Environ())
+	return parseAdditionalHeadersFromEnviron(os.Environ())
 }
 
 const additionalHeaderPrefix = "SRC_HEADER_"
+const additionalHeadersKey = "SRC_HEADERS"
 
-func parseAdditionalHeadersFromMap(environ []string) map[string]string {
+func parseAdditionalHeadersFromEnviron(environ []string) map[string]string {
 	additionalHeaders := map[string]string{}
 	for _, value := range environ {
 		parts := strings.SplitN(value, "=", 2)
 		if len(parts) != 2 {
 			continue
 		}
 
+		if parts[0] == additionalHeadersKey && parts[1] != "" {
+			// This regex removes all leading and trailing spaces from the environment variable. We need to do this
+			// because most shells add quotes to a string when it contains a new line. Tested with `bash, fish & zsh`
+			// and they all have the same behavior.
+			re := regexp.MustCompile(`^"|"$`)
+			headers := re.ReplaceAllString(parts[1], "")
+			// Most shell convert line breaks added to a string, so we need to replace all occurence of the stringified line
+			// breaks to actual line breaks.
+			headers = strings.ReplaceAll(headers, `\n`, "\n")
+			splitHeaders := strings.Split(headers, "\n")
+
+			for _, h := range splitHeaders {
+				p := strings.SplitN(h, ":", 2)
+				if len(parts) != 2 || p[1] == "" {
+					continue
+				}
+
+				additionalHeaders[strings.ToLower(p[0])] = strings.Trim(p[1], " ")
```

Comment:
We should probably trim `p[0]` as well in case somebody gives us something like `Authorization : foo bar`. (Some people really like spaces.)